File: day11/main.ml
#load "str.cma";;

let rec count_stones a blinks =
    if blinks == 0 then
        1
    else
        if a == 0 then
            count_stones 1 (blinks-1)
        else
            let s = string_of_int a in
            let l = String.length s in
            if (l mod 2) == 0 then
                let l_stone = int_of_string (String.sub s 0 (l/2)) in
                let r_stone = int_of_string (String.sub s (l/2) (l/2)) in
                (count_stones l_stone (blinks-1)) + (count_stones r_stone (blinks-1))
            else
                count_stones (a*2024) (blinks-1)

let () =
    let blinks = int_of_string (Sys.argv.(1)) in
    let filepath = "input" in
    let input_channel = open_in filepath in 
    let line = input_line input_channel in
    let stones = List.map int_of_string (Str.split (Str.regexp " ") line) in
    let f acc a =
        acc + (count_stones a blinks)
    in
    let result = List.fold_left f 0 stones in
    Printf.printf "Result after %d blinks: %d\n" blinks result

File: day11/main.py
import functools
import sys


@functools.cache
def count_stones(s, b):
    if b == 0:
        return 1
    if s == 0:
        return count_stones(1, b-1)
    else:
        s_str = str(s)
        l = len(s_str)
        if l%2 == 0:
            l_s = int(s_str[:l//2])
            r_s = int(s_str[l//2:])
            return count_stones(l_s, b-1) + count_stones(r_s, b-1)
        else:
            return count_stones(s*2024, b-1)

filepath = sys.argv[1]
blinks = int(sys.argv[2])

with open(filepath, "r") as f:
    result = sum(count_stones(s, blinks) for s in map(int, f.read().strip().split(" ")))
print(f"Number of stones after {blinks} blinks: {result}")

